File: src/clients/lib/xmmsclient++/playback.cpp
/*  XMMS2 - X Music Multiplexer System
 *  Copyright (C) 2003-2007 XMMS2 Team
 *
 *  PLUGINS ARE NOT CONSIDERED TO BE DERIVED WORK !!!
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 */

#include <xmmsclient/xmmsclient.h>
#include <xmmsclient/xmmsclient++/client.h>
#include <xmmsclient/xmmsclient++/playback.h>
#include <xmmsclient/xmmsclient++/signal.h>
#include <xmmsclient/xmmsclient++/helpers.h>
#include <xmmsclient/xmmsclient++/mainloop.h>
#include <xmmsclient/xmmsclient++/typedefs.h>
#include <xmmsclient/xmmsclient++/dict.h>

#include <boost/bind.hpp>

#include <iostream>
#include <string>

namespace Xmms
{
	
	Playback::~Playback()
	{
	}

	VoidResult Playback::tickle() const
	{
		xmmsc_result_t* res =
			call( connected_, boost::bind( xmmsc_playback_tickle, conn_ ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::stop() const
	{
		xmmsc_result_t* res =
			call( connected_, boost::bind( xmmsc_playback_stop, conn_ ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::pause() const
	{
		xmmsc_result_t* res =
			call( connected_, boost::bind( xmmsc_playback_pause, conn_ ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::start() const
	{
		xmmsc_result_t* res =
			call( connected_, boost::bind( xmmsc_playback_start, conn_ ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::seekMs(unsigned int milliseconds) const
	{
		xmmsc_result_t* res =
			call( connected_,
			      boost::bind( xmmsc_playback_seek_ms, conn_, milliseconds ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::seekMsRel(int milliseconds) const
	{
		xmmsc_result_t* res =
			call( connected_,
			      boost::bind( xmmsc_playback_seek_ms_rel,
			                   conn_, milliseconds ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::seekSamples(unsigned int samples) const
	{
		xmmsc_result_t* res =
			call( connected_,
			      boost::bind( xmmsc_playback_seek_samples, conn_, samples ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playback::seekSamplesRel(int samples) const
	{
		xmmsc_result_t* res =
			call( connected_,
			      boost::bind( xmmsc_playback_seek_samples_rel,
			                   conn_, samples ) );
		return VoidResult( res, ml_ );
	}

	UintResult Playback::currentID() const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playback_current_id, conn_ ) );
		return UintResult( res, ml_ );
	}

	StatusResult Playback::getStatus() const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playback_status, conn_ ) );

		return StatusResult( res, ml_ );
	}

	UintResult Playback::getPlaytime() const
	{
		xmmsc_result_t* res = 
		    call( connected_, boost::bind( xmmsc_playback_playtime, conn_ ) );
		return UintResult( res, ml_ );
	}

	VoidResult Playback::volumeSet(const std::string& channel,
	                               unsigned int volume) const
	{
		xmmsc_result_t* res =
			call( connected_,
			      boost::bind( xmmsc_playback_volume_set, conn_,
			                   channel.c_str(), volume ) );
		return VoidResult( res, ml_ );
	}

	DictResult Playback::volumeGet() const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playback_volume_get, conn_ ) );
		return DictResult( res, ml_ );
	}

	Playback::Playback( xmmsc_connection_t*& conn, bool& connected,
	                    MainloopInterface*& ml ) :
		conn_( conn ), connected_( connected ), ml_( ml )
	{
	}

}

File: src/clients/lib/dotnet/Playback.cs
//
//  .NET bindings for the XMMS2 client library
//
//  Copyright (C) 2005 Daniel Svensson, <nano@nittioonio.nu>
//  Copyright (C) 2008 Tilman Sauerbeck, <tilman@xmms.org>
//
//  This library is free software; you can redistribute it and/or
//  modify it under the terms of the GNU Lesser General Public
//  License as published by the Free Software Foundation; either
//  version 2.1 of the License, or (at your option) any later version.
//
//  This library is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
//  Lesser General Public License for more details.
//

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Xmms.Client {

	public class Playback {
		public Playback(Client c) {
			client = c;
		}

		public Result Stop() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_stop(client.Connection);

			return new Result(resultHandle);
		}

		public Result Tickle() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_tickle(client.Connection);

			return new Result(resultHandle);
		}

		public Result Start() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_start(
					client.Connection
				);

			return new Result(resultHandle);
		}

		public Result Pause() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_pause(
					client.Connection
				);

			return new Result(resultHandle);
		}

		public Result<Value.UInt32> CurrentID() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_current_id(
					client.Connection
				);

			return new Result<Value.UInt32>(resultHandle);
		}

		public Result SeekMs(uint ms) {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_seek_ms(
					client.Connection, ms
				);

			return new Result(resultHandle);
		}

		public Result SeekMsRel(uint ms) {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_seek_ms_rel(
					client.Connection, ms
				);

			return new Result(resultHandle);
		}

		public Result SeekSamples(uint samples) {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_seek_samples(
					client.Connection, samples
				);

			return new Result(resultHandle);
		}

		public Result SeekSamplesRel(uint samples) {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_seek_samples_rel(
					client.Connection, samples
				);

			return new Result(resultHandle);
		}

		public Result<Value.UInt32> Playtime() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_playtime(
					client.Connection
				);

			return new Result<Value.UInt32>(resultHandle);
		}

#if false
		public PlaybackStatus Status() {
			return (PlaybackStatus) Helper.Sync<uint>(NativeMethods.xmmsc_playback_status, client);
		}

		public void Status(Callback<PlaybackStatus> cb) {
			Helper.ASync<Xmms.PlaybackStatus>(NativeMethods.xmmsc_playback_status, client, cb);
		}
#endif

		public Result VolumeSet(
			string channelName, uint channelVolume
		) {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_volume_set(
					client.Connection,
					channelName, channelVolume
				);

			return new Result(resultHandle);
		}

		public Result<Value.Dictionary<Value.UInt32>> VolumeGet() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_playback_volume_get(
					client.Connection
				);

			return new Result<Value.Dictionary<Value.UInt32>>(
				resultHandle
			);
		}

#if false
		public void BroadcastVolumeChanged(Callback<uint> cb) {
			Helper.ASync<uint>(NativeMethods.xmmsc_broadcast_playback_volume_changed, client, cb);
		}

		public void BroadcastStatus(Callback<PlaybackStatus> cb) {
			Helper.ASync<Xmms.PlaybackStatus>(NativeMethods.xmmsc_broadcast_playback_status, client, cb);
		}
#endif

		public BroadcastResult<Value.UInt32> BroadcastCurrentID() {
			ResultHandle resultHandle =
				NativeMethods.xmmsc_broadcast_playback_current_id(
					client.Connection
				);

			return new BroadcastResult<Value.UInt32>(resultHandle);
		}

#if false
		public void SignalPlaytime (Xmms.Callback<uint> cb)
		{
			Helper.ASync<uint>(NativeMethods.xmmsc_signal_playback_playtime, client, cb);
		}
#endif

		private readonly Client client;
	}
}
